### ISSUE TITLE
Develop main simplification function

### ISSUE THREAD

**Issue by martinnormark** [martinnormark/neural-mesh-simplification#27]
Where we are:
1. We have implemented and tested the PointSampler module.
2. We have implemented and tested the EdgePredictor module.
3. We have just completed and tested the FaceClassifier module.

These three components form the core of the Neural Mesh Simplification model as described in the paper.

Next steps:
The primary task now is to integrate these modules into the main NeuralMeshSimplification class. Here's an outline of what this integration might look like:

1. Implement the NeuralMeshSimplification class:

```python
class NeuralMeshSimplification(nn.Module):
    def __init__(self, input_dim, hidden_dim, num_layers=3, k=20):
        super(NeuralMeshSimplification, self).__init__()
        self.point_sampler = PointSampler(input_dim, hidden_dim)
        self.edge_predictor = EdgePredictor(hidden_dim, hidden_dim)
        self.face_classifier = FaceClassifier(hidden_dim, hidden_dim, num_layers, k)

    def forward(self, x, pos, batch=None):
        # Sample points
        sampled_indices, sampled_probs = self.point_sampler(x, pos, batch)
        sampled_x = x[sampled_indices]
        sampled_pos = pos[sampled_indices]
        
        # Predict edges
        edge_index, edge_probs = self.edge_predictor(sampled_x, sampled_pos, batch)
        
        # Generate candidate triangles
        candidate_triangles = self.generate_candidate_triangles(edge_index, edge_probs)
        
        # Classify faces
        face_probs = self.face_classifier(sampled_x, sampled_pos, batch)
        
        return sampled_indices, sampled_probs, edge_index, edge_probs, candidate_triangles, face_probs

    def generate_candidate_triangles(self, edge_index, edge_probs):
        # Implement logic to generate candidate triangles from edges
        pass
```

2. Implement the generate_candidate_triangles method:
   This method should take the predicted edges and their probabilities and generate a set of candidate triangles. The exact implementation will depend on how you want to form triangles from the edges.

3. Implement loss functions:
   As described in the paper, you'll need to implement several loss functions, including:
   - Probabilistic Chamfer Distance
   - Probabilistic Surfaces Distance
   - Triangle Collision
   - Edge Crossings
   - Overlapping triangles

4. Create a training loop:
   Implement a training script that uses these loss functions and trains the NeuralMeshSimplification model on your dataset.

5. Implement evaluation metrics:
   Create functions to evaluate the quality of the simplified meshes, possibly including:
   - Chamfer distance
   - Normal consistency
   - Edge preservation

6. Testing and validation:
   Create comprehensive tests for the NeuralMeshSimplification class and its methods.

7. Optimization and refinement:
   Once the basic implementation is working, you may want to optimize for performance and refine the model based on experimental results.

The next immediate step would be to implement the NeuralMeshSimplification class as outlined above, and then move on to implementing the generate_candidate_triangles method. After that, you can focus on the loss functions and training loop.
